nhrp, ospf: add nhrp multicast for OSPF DMVPN

reubendowle opened the pull request:
This pull requests adds nhrp multicast support, and enables the use of OSPF within a DMVPN network.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/fb075323e94974d7328bc019c5c8e405/raw/7c670b1e7144d772f24ea750b11408cf7aa7ecf1/cr_8153_1614291670.diff | git apply
```

</p>

<p>

```diff
diff --git a/ospfd/ospf_lsa.c b/ospfd/ospf_lsa.c
index da779c9ea..5ab584daf 100644
--- a/ospfd/ospf_lsa.c
+++ b/ospfd/ospf_lsa.c
@@ -488,7 +488,8 @@ static int lsa_link_ptop_set(struct stream **s, struct ospf_interface *oi)
 	if ((nbr = ospf_nbr_lookup_ptop(oi)))
 		if (nbr->state == NSM_Full) {
 			if (CHECK_FLAG(oi->connected->flags,
-				       ZEBRA_IFA_UNNUMBERED) && !oi->ptp_dmvpn) {
+				       ZEBRA_IFA_UNNUMBERED)
+			    && !oi->ptp_dmvpn) {
 				/* For unnumbered point-to-point networks, the
 				   Link Data field
 				   should specify the interface's MIB-II ifIndex
@@ -506,7 +507,8 @@ static int lsa_link_ptop_set(struct stream **s, struct ospf_interface *oi)
 		}
 
 	/* no need for a stub link for unnumbered interfaces */
-	if (oi->ptp_dmvpn || !CHECK_FLAG(oi->connected->flags, ZEBRA_IFA_UNNUMBERED)) {
+	if (oi->ptp_dmvpn
+	    || !CHECK_FLAG(oi->connected->flags, ZEBRA_IFA_UNNUMBERED)) {
 		/* Regardless of the state of the neighboring router, we must
 		   add a Type 3 link (stub network).
 		   N.B. Options 1 & 2 share basically the same logic. */
diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 8199eb67e..5307da1b6 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -8346,17 +8346,16 @@ DEFUN_HIDDEN (no_ospf_hello_interval,
 	return no_ip_ospf_hello_interval(self, vty, argc, argv);
 }
 
-DEFUN (ip_ospf_network,
-       ip_ospf_network_cmd,
-       "ip ospf network <broadcast|non-broadcast|point-to-multipoint|point-to-point [dmvpn]>",
-       "IP Information\n"
-       "OSPF interface commands\n"
-       "Network type\n"
-       "Specify OSPF broadcast multi-access network\n"
-       "Specify OSPF NBMA network\n"
-       "Specify OSPF point-to-multipoint network\n"
-       "Specify OSPF point-to-point network\n"
-       "Specify OSPF point-to-point DMVPN network\n")
+DEFUN(ip_ospf_network, ip_ospf_network_cmd,
+      "ip ospf network <broadcast|non-broadcast|point-to-multipoint|point-to-point [dmvpn]>",
+      "IP Information\n"
+      "OSPF interface commands\n"
+      "Network type\n"
+      "Specify OSPF broadcast multi-access network\n"
+      "Specify OSPF NBMA network\n"
+      "Specify OSPF point-to-multipoint network\n"
+      "Specify OSPF point-to-point network\n"
+      "Specify OSPF point-to-point DMVPN network\n")
 {
 	VTY_DECLVAR_CONTEXT(interface, ifp);
 	int idx = 0;
@@ -8385,7 +8384,7 @@ DEFUN (ip_ospf_network,
 	}
 
 	if (IF_DEF_PARAMS(ifp)->type == old_type
-		&& IF_DEF_PARAMS(ifp)->ptp_dmvpn == old_ptp_dmvpn)
+	    && IF_DEF_PARAMS(ifp)->ptp_dmvpn == old_ptp_dmvpn)
 		return CMD_SUCCESS;
 
 	SET_IF_PARAM(IF_DEF_PARAMS(ifp), type);
@@ -11583,8 +11582,9 @@ static int config_write_interface_one(struct vty *vty, struct vrf *vrf)
 					vty_out(vty, " ip ospf network %s",
 						ospf_int_type_str
 							[params->type]);
-					if (params->type == OSPF_IFTYPE_POINTOPOINT
-						&& params->ptp_dmvpn)
+					if (params->type
+						    == OSPF_IFTYPE_POINTOPOINT
+					    && params->ptp_dmvpn)
 						vty_out(vty, " dmvpn");
 					if (params != IF_DEF_PARAMS(ifp) && rn)
 						vty_out(vty, " %pI4",

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

Jafaral commented:
fixes #6276

Jafaral commented:
@reubendowle Thank you for this excellent contribution! Can  you please fix the build warnings ?

donaldsharp commented:
doc changes and topotest?

reubendowle commented:
> 
> 
> doc changes and topotest?

The new commands are already documented as part of the pull request. Is that sufficient?

I am not familiar with the topotests. How would I go about adding that?

Jafaral commented:
We don't have any topotests for nhrpm but we can start somewhere. Example for other protocols :

https://github.com/FRRouting/frr/tree/master/tests/topotests

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

pguibert6WIND commented on nhrpd/nhrp_multicast.c:
```diff
@@ -47,19 +47,22 @@ static void nhrp_multicast_send(struct nhrp_peer *p, struct zbuf *zb)
-		   ret, zbuf_used(zb), addrlen);
+			 sockunion_get_addr(&p->vc->remote.nbma), addrlen,
+			 addrlen == 4 ? 0x0800 : 0x86DD);
+
```
it should be the good time to add if_ether.h in 
https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/if_ether.h#L54, in include/linux/if_ether.h

Alternatively, just add on the header definition of ETH_P_IPV6 ( I see that ETH_ALEN is also redefined in prefix.h, and adding if_ether.h may be more complicated than expected)

reubendowle replied:
linux/if_ether.h is already included via netinet/if_ether.h. So very simple to change code to use defines instead of hard-coded values.

pguibert6WIND reviewed (changes requested):
Hi,
nice changes.
I have some comments however that should be addresses.

About ospfv3, did you try to play with it too to test ipv6 protocol multicast addresses ?

reubendowle commented on ospfd/ospf_vty.c:
```diff
@@ -8346,20 +8346,21 @@ DEFUN_HIDDEN (no_ospf_hello_interval,
+DEFUN(ip_ospf_network, ip_ospf_network_cmd,
+      "ip ospf network <broadcast|non-broadcast|point-to-multipoint|point-to-point [dmvpn]>",
+      "IP Information\n"
+      "OSPF interface commands\n"
```
@pguibert6WIND, I cannot see any existing yang model for nhrp

reubendowle replied:
I had a quick look at the existing yang model, and it is not immediately clear how to add this in. I am not familiar with yang, and I don't know how I would test any changes I made with this.

pguibert6WIND replied:
I apologise. actually, ospf is not ready yet for migrating to northbound.
however, to clarify: what kind of network should we configure when we deal with ospf shortcuts ? are we still in 'ip ospf ptp dmvpn' ?

Answering this will help clarify when we move to yang model. one possiblity would be to make a leaf independent, that could be set whatever the ospf network type.

+      leaf dmvpn {
+        type boolean;
+        default "false";
+        description
+          "Used with point-to-point networks.";
+      }
+

reubendowle replied:
When using OSPF with DMVPN you need to set it up as a point-to-point at a spoke. However the hub will be setup as point-to-multipoint, and so this would not normally work. The 'point-to-point dmvpn' option makes a spoke send packets that will work correctly in this scenario.

I believe this is actually a subtle violation of the RFC for OSPF, but it is needed to make this work.

Note also that Cisco routers don't have this issue, because in that case you configure the GRE tunnel with a subnet, and setup ospf as 'network broadcast' at both hub and spoke. However in nhrpd we must set the GRE tunnel to a /32 point-point configuration.

pguibert6WIND replied:
yang changes will be  done in separate pull request

reubendowle commented on nhrpd/nhrp_multicast.c:
```diff
@@ -0,0 +1,307 @@
+
+	n = znl_nlmsg_push(zb, (NFNL_SUBSYS_ULOG << 8) | NFULNL_MSG_CONFIG,
+			   NLM_F_REQUEST | NLM_F_ACK);
+	nf = znl_push(zb, sizeof(*nf));
```
I am not sure NF_ENQUEUE would end up being much more efficient. The difference for such low-rate packets is not going to be worth worrying about.

I have updated documentation to include how to drop the original multicast packet.
Note that when this block is put in place this causes the sendmsg in ospfd to report failure, so you get this log entry created:
ospfd: [EC 100663299] *** sendmsg in ospf_write failed to 224.0.0.5, id 46113, off 0, len 64, interface abc, mtu 1500: Operation not permitted

I am not sure I follow you about the ptp tunnel. The GRE tunnel will always encapsulate the multicast and try to send it.

I also don't know enough about any possible kernel implementation to be able to comment on this.

pguibert6WIND replied:
nhrp has the ability to work on a point to point interface.
this is the static nhrp configuration
that means that tunnel destination has a correct value, and then that original multicast packets may not be dropped.
nhrp static config can look like below on  a tunnel with tunnel destination set to 44.44.44.44.
```
vtysh(config-iface-gre1)# ip nhrp-map 10.255.255.4 nbma 44.44.44.44
```

so my question is if nhrp code can work like to ptomp tunnels, or should there be a hack in the code to check for ptop or ptomp tunnels ? (in order to simplify the configuration on the spoke side).

reubendowle replied:
All the testing I have done has been on multi point tunnels, and in that situation there is no need to specify an 'ip nhrp map'. I am not really sure I understand a use-case for static nhrp, as that looses the big advantage of DMVPN where routing etc gets auto discovered.

I am also not sure how this is relevant to if multicast is dropped or not. I think the original multicast packet would be sent on GRE tunnel regardless of if it is point-point or point-multi-point. I know from my testing with multi-point setups that the outer IP header ends up getting the multicast address as well, which means that the packet obviously will never reach its destination. Perhaps with point-point GRE the multicast packet would be sent to destination, although I never checked. If that were the case, you could just not configure any map multicast.

Jafaral replied:
> I am not sure NF_ENQUEUE would end up being much more efficient. The difference for such low-rate packets is not going to be worth worrying about.
> 
 This is true for OSPF protocol multicast but isn't this supposed to  work with all multicast streams  as long as they hit the interface? So you could get high bandwidth here, right?

Jafaral replied:
> All the testing I have done has been on multi point tunnels, and in that situation there is no need to specify an 'ip nhrp map'. I am not really sure I understand a use-case for static nhrp, as that looses the big advantage of DMVPN where routing etc gets auto discovered.

While this is generally true, it is still useful in some cases to have static maps such as testing or when you have a situation where you have a predefined set of nhrp neighbors. Cisco do have such  command, don't they?

reubendowle replied:
> 
> This is true for OSPF protocol multicast but isn't this supposed to work with all multicast streams as long as they hit the interface? So you could get high bandwidth here, right?

The recommended NFLOG redirection in the documentation is only for the 224.0.0/24 local network control block, as that will capture all of the routing protocols. I think it is unlikely you would see high traffic on this multicast range.

That said, you are right that there is no reason you could not redirect any multicast, so you could encounter high bandwidth. But I don't really have the time to re-engineer this. Perhaps a note in the documentation about potential performance issues with high rate traffic would be ok for now. If someone else wants to pick it up and look into optimising that they can.

reubendowle replied:
> 
> While this is generally true, it is still useful in some cases to have static maps such as testing or when you have a situation where you have a predefined set of nhrp neighbors. Cisco do have such command, don't they?

Yes, cisco have an equivalent command. I actually think that for a point-point tunnel you would not actually need to use map multicast, as the point-point tunnel will send the multicast through already. Cisco have a note to this effect in their documentation, and I think the same is true for a GRE point-point tunnel in linux. I will do a quick test tomorrow and check if this is the case, and if so I will change the documentation to reflect this.

Jafaral replied:
@reubendowle  did you get a chance to do the test ?

reubendowle commented:
> 
> About ospfv3, did you try to play with it too to test ipv6 protocol multicast addresses ?

No attempt has been made to test any multicast protocols other than ospfv2.

pguibert6WIND commented on ospfd/ospf_lsa.c:
```diff
@@ -469,6 +469,12 @@ char link_info_set(struct stream **s, struct in_addr id, struct in_addr data,
+ * hub then expects this router to populate the stub network and also Link Data
+ * Field set to IP Address and not MIB-II ifIndex
+ */
 static int lsa_link_ptop_set(struct stream **s, struct ospf_interface *oi)
```
even if I do a shortcut, do we still consider spoke as point to point ?

reubendowle replied:
Yes. There is normally no map multicast done for the shortcuts. Spoke was OSPF only communicating with the hub in my testing.

Now I think in principle you could setup the spokes with map multicast dynamic, and then have associations with multiple other routers. Although I did not test this out, and I am not sure how much sense this would make. The spokes learn to do shortcuts through traffic indications from the base.

amollad replied:
> even if I do a shortcut, do we still consider spoke as point to point ?

Yes, spoke shall still be point-to-point. Note that "point-to-point" is at the OSPF layer which means spoke forms OSPF neighbor adjacency ONLY with hub. Thus, even if spoke-spoke shortcut is established, they won't become OSPF neighbors (as OSPF multicasts are only getting exchanged between hub and spoke)

However, there was one comment in cumulus network doc which said

"OSPF unnumbered is supported with point-to-point interfaces **only** and does not support network statements."

https://docs.cumulusnetworks.com/cumulus-linux-43/Layer-3/OSPF/Open-Shortest-Path-First-v2-OSPFv2/##ospfv2-unnumbered

But, if FRR OSPF point-to-multipoint network type supports unnumbered interfaces without any restrictions then I don't see a reason why we cannot make "ip ospf network point-to-multipoint" work on spoke

pguibert6WIND replied:
I did not find a literature saying it is forbidden to do adjacencies with shortcuts ( neither the opposite:)).
I am ok for keeping the code as it is today.

However, to better understand what can be possible, if we agree that it is possible to configure ptomultipoint, to establish adjacencies with shortcuts, then I think there should be a change in the function `nhrp_multicast_foreach()', where each dynamic cache entry would be sent multicast traffic ?

And if we use a dynamic multicast list based on the list of cache entries, which dynamic entries should we choose ?
As a note on [https://github.com/FRRouting/frr/pull/7747/commits/ac95bcefe4492e620c33ab17b2c92c203abc2a5b], I wonder if it is really feasible.

note that I encountered 

```
show ip nhrp cache 
Iface    Type     Protocol                 NBMA                     Flags  Identity
gre1     local    11.255.255.2             -                               -
gre1     nhs      11.255.255.1             10.125.0.1                T    
gre1     dynamic  11.150.0.1               10.125.0.4                TA   
gre1     dynamic  11.255.255.4             10.125.0.4                TA   
```

amollad replied:
I'm not sure, I followed your suggestion regarding change in 'nhrp_multicast_foreach()'. If you see 'nhrp_multicast_forward()' - this function sends multicast to each dynamic cache entry if "ip nhrp map multicast dynamic" is configured. Do you see any issues there?

amollad replied:
> gre1     dynamic  11.150.0.1               10.125.0.4                TA   
> gre1     dynamic  11.255.255.4             10.125.0.4                TA

How did you land into this?

I have only seen this happening (i.e. different protocol addresses but same NBMA) when both the devices are behind the same NAT Router. This use case is not supported (Cisco also does not supports this) - i.e. we cannot have multiple spokes behind a same NAT router

pguibert6WIND replied:
> > gre1     dynamic  11.150.0.1               10.125.0.4                TA
> > gre1     dynamic  11.255.255.4             10.125.0.4                TA
> 
> How did you land into this?
> 
> I have only seen this happening (i.e. different protocol addresses but same NBMA) when both the devices are behind 

Possibly, this is related to the branch used for testing.
I will retest ac95bce with master branch the nhrp hub use case. and will let you know.

Now I have the confirmation that it should not be allowed.
Thanks

pguibert6WIND replied:
the issue is related to master branch.
there are high chances that recent modification  brought on master broken the master branch:
```
north-vm# show ip nhrp shortcut                                                                   |
Type     Prefix                   Via                      Identity                               |
dynamic  11.150.0.0/24            11.255.255.4                                                    |
north-vm# show ip nhrp cache                                                                      |
Iface    Type     Protocol                 NBMA                     Flags  Identity               |
gre1     local    11.255.255.2             -                               -                      |
gre1     nhs      11.255.255.1             10.125.0.1               UT                            |
gre1     dynamic  11.150.0.1               10.125.0.4               UT                            |
gre1     dynamic  11.255.255.4             10.125.0.4               UT                            |
north-vm#  
```

amollad replied:
@pguibert6WIND Aah, now I understand what you are doing. I did not realized you are trying shortcut. So what you see above is correct output on spoke
11.255.255.4 is the protocol address and 11.150.0.0/24 is subnet you are creating shortcut for.

You should also see "sh ip route nhrp" and you will see three route entries, something like below:

N>* 11.255.255.4/32 [10/0] is directly connected, gre1, weight 1, 00:00:17
N>* 11.150.0.0/24 [10/0] via 11.255.255.4, gre1 onlink, weight 1, 00:00:17
N>* 11.150.0.1/32 [10/0] is directly connected, gre1, weight 1, 00:00:17

You can also verify "ip neigh show" command on linux and you will find neighbor entries for shortcut are correct

**So basically, I do not see anything wrong in above output you posted. That is correct on spoke.**

amollad replied:
One configuration we've not tested is "ip nhrp map multicast dynamic" on spokes. What we have tested is below:

Hub:  ip nhrp map multicast dynamic
spokes: ip nhrp map multicast X.X.X.X

If we've to support "ip nhrp map mutlicast dynamic" on spokes then we will need to add a flag in 'struct nhrp_cache' indicating that the dynamic entry is for "protocol address" and not for "external subnet". By 'protocol address', I mean spoke's tunnel source interface IP address (11.255.255.4 in your case I assume)

Then, 'nhrp_multicast_forward()' can forward multicast to only those dynamic cache entries which have this 'protocol address' flag set.

Above can be done I believe but this can be taken as enhancement as I don't think many dmvpn installations will use "ip nhrp map multicast dynamic" on spokes

pguibert6WIND replied:
I agree @amollad. thanks for feedback. subject closed.

pguibert6WIND commented on nhrpd/nhrp_multicast.c:
```diff
@@ -0,0 +1,307 @@
+	struct mcast_ctx ctx;
+
+	debugf(NHRP_DEBUG_COMMON,"Inside %s\n", __func__);
+
```
some message more relevant:
"received multicast packet size %u"

reubendowle replied:
I have fixed this up, and also fixed a possible reference to uninitialized variable I noticed in the code when I looked at it again.

pguibert6WIND commented on doc/user/nhrpd.rst:
```diff
@@ -180,6 +180,34 @@ https://git-old.alpinelinux.org/user/tteras/strongswan/
+
+.. index::  nhrp multicast-nflog-group (1-65535)
+.. clicmd:: nhrp multicast-nflog-group (1-65535)
+
```
that command is not implemented.

reubendowle replied:
I missed the change @amollad did to implement that command. I have added that change onto this pull request now.

pguibert6WIND reviewed (changes requested):
it looks like command:
nhrp multicast-nflog-group
is not yet implemented.

can you fix this, please?

reubendowle commented:
> 
> # Continuous Integration Result: FAILED
> 
> See below for issues.
> CI System Testrun URL: https://ci1.netdef.org/browse/FRR-FRRPULLREQ-17727/
> 
> This is a comment from an automated CI system.
> For questions and feedback in regards to this CI system, please feel free to email
> Martin Winter - mwinter (at) opensourcerouting.org.
> ### Get source / Pull Request: Successful
> ### Building Stage: Successful
> ### Basic Tests: Failed
> Topotests Ubuntu 16.04 amd64 part 6: Failed (click for details)
> Topotests Ubuntu 18.04 arm8 part 7: Failed (click for details)
> Topotests Ubuntu 18.04 arm8 part 7: No useful log found
> Topotests Ubuntu 16.04 i386 part 6: Failed (click for details)
> Topotests Ubuntu 16.04 i386 part 7: Failed (click for details)
> Successful on other platforms/tests
> ### Warnings Generated during build:
> Checkout code: Successful with additional warnings

pguibert6WIND commented on nhrpd/nhrp_multicast.c:
```diff
@@ -130,21 +127,18 @@ static void netlink_mcast_log_handler(struct nlmsghdr *msg, struct zbuf *zb)
-	};
+	debugf(NHRP_DEBUG_COMMON, "Received multicast packet on %s len %zu\n",
+		   ctx.ifp->name, zbuf_used(ctx.pkt));
 
```
- \n is not necessary.
- also the comment may be more explicit. I did not realize at first, but we need to understand that this is a packet going out that is intercepted ( and not incoming ospf messages).
what about mesage like:
"Intercepted multicast packet leaving %s len %zu"

pguibert6WIND replied:
missing this comment ?

pguibert6WIND replied:
thanks

pguibert6WIND reviewed (changes requested):
I encountered an unexplained error on hub configuration when receiving OSPF packet, I have following message
```
2021/03/17 08:48:32 OSPF: [EC 134217741] interface gre1:11.255.255.1: ospf_read network address is not same [11.255.255.2]
2021/03/17 08:48:32 OSPF: ospf_read[11.255.255.2]: Header check failed, dropping.
2021/03/17 08:48:32 OSPF: Hello sent to [224.0.0.5] via [ntfp1:11.175.0.1].
```
I suppose there is something wrong:
- at best,  the config may be incomplete.
- at worse, ospf code is incomplete.
below is ospf and nhrp configuration.
can you please have a look please?

```
nhrp nflog-group 6
nhrp multicast-nflog-group 2
!
interface gre1
 ip nhrp map 11.255.255.1 local
 ip nhrp map multicast dynamic
 ip nhrp network-id 42
 ip nhrp redirect
!
interface gre1
 ip ospf area 0
 ip ospf dead-interval 10
 ip ospf hello-interval 2
 ip ospf network point-to-multipoint
!
```

pguibert6WIND commented on doc/user/nhrpd.rst:
```diff
@@ -189,11 +189,14 @@ nhrpd can be configured to forward multicast packets, allowing routing
-   iptables -I OUTPUT -d 224.0.0.0/24 -o gre1 -j NFLOG --nflog-group 2
+   iptables -A OUTPUT -d 224.0.0.0/24 -o gre1 -j NFLOG --nflog-group 2
+   iptables -A OUTPUT -d 224.0.0.0/24 -o gre1 -j DROP
 
```
about this error message:
```
2021/03/17 08:48:32 OSPF: [EC 100663299] *** sendmsg in ospf_write failed to 224.0.0.5, id 46163, off 0, len 64, interface gre1, mtu 1452: Operation not permitted
```
it seems the error is derived from the DROP rule.
having flog_err messages should absolutely be avoided.
is it possible to detect errno "operation not permitted" to avoid that error message when sending ospf packets?

reubendowle replied:
I have put an extra change so that this failure message will be suppressed.

pguibert6WIND reviewed (changes requested):
some more comments; and this will be ok.

amollad commented:
> I encountered an unexplained error on hub configuration when receiving OSPF packet, I have following message
> 
> ```
> 2021/03/17 08:48:32 OSPF: [EC 134217741] interface gre1:11.255.255.1: ospf_read network address is not same [11.255.255.2]
> 2021/03/17 08:48:32 OSPF: ospf_read[11.255.255.2]: Header check failed, dropping.
> 2021/03/17 08:48:32 OSPF: Hello sent to [224.0.0.5] via [ntfp1:11.175.0.1].
> ```

Yes, one commit has been missed out in the pull request - there is a small change to ignore network mask in point-to-multipoint ospf network type when interface is configured with /32 prefix. I'll ask @reubendowle to push the commit

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

reubendowle commented:
> 
> 
> > I encountered an unexplained error on hub configuration when receiving OSPF packet, I have following message
> > ```
> > 2021/03/17 08:48:32 OSPF: [EC 134217741] interface gre1:11.255.255.1: ospf_read network address is not same [11.255.255.2]
> > 2021/03/17 08:48:32 OSPF: ospf_read[11.255.255.2]: Header check failed, dropping.
> > 2021/03/17 08:48:32 OSPF: Hello sent to [224.0.0.5] via [ntfp1:11.175.0.1].
> > ```
> 
> Yes, one commit has been missed out in the pull request - there is a small change to ignore network mask in point-to-multipoint ospf network type when interface is configured with /32 prefix. I'll ask @reubendowle to push the commit

I have pushed the change. I have not tested this on my end - @amollad, can you confirm this is the changes you expected there and that it works.

pguibert6WIND reviewed (changes requested):
> > > I encountered an unexplained error on hub configuration when receiving OSPF packet, I have following message
> > > ```
> > > 2021/03/17 08:48:32 OSPF: [EC 134217741] interface gre1:11.255.255.1: ospf_read network address is not same [11.255.255.2]
> > > 2021/03/17 08:48:32 OSPF: ospf_read[11.255.255.2]: Header check failed, dropping.
> > > 2021/03/17 08:48:32 OSPF: Hello sent to [224.0.0.5] via [ntfp1:11.175.0.1].
> > > ```
> > 
> > 
> > Yes, one commit has been missed out in the pull request - there is a small change to ignore network mask in point-to-multipoint ospf network type when interface is configured with /32 prefix. I'll ask @reubendowle to push the commit
> 
> I have pushed the change. I have not tested this on my end - @amollad, can you confirm this is the changes you expected there and that it works.

I tested it and this does not change the behaviour.
the issue takes place on hub, when receiving OSPF packets.
```
2021/03/18 09:03:27 OSPF: ospf_read[11.255.255.4]: Header check failed, dropping.
```

maybe you did not face this issue because you use an other device as hub.
On my side, I use frrouting as hub, and I expect it to work.


+ some minor comments remaining.

amollad commented:
> 
> I tested it and this does not change the behaviour.
> the issue takes place on hub, when receiving OSPF packets.

We've tested with FRR as hub also. Can you please tell what header check is failing?

amollad commented on ospfd/ospf_packet.c:
```diff
@@ -2427,6 +2430,11 @@ static int ospf_check_network_mask(struct ospf_interface *oi,
 	    || oi->type == OSPF_IFTYPE_VIRTUALLINK)
 		return 1;
 
+	/* Ignore mask check for max prefix length (32) */
```
@pguibert6WIND this change is added to ignore network mask check for interface is configured with /32 prefix

amollad commented on ospfd/ospf_packet.c:
```diff
@@ -2427,6 +2430,11 @@ static int ospf_check_network_mask(struct ospf_interface *oi,
+	if (oi->type == OSPF_IFTYPE_POINTOMULTIPOINT
+	    && oi->address->prefixlen == IPV4_MAX_BITLEN)
+		return 1;
+
```
@pguibert6WIND this change is added to ignore network mask check for interface is configured with /32 prefix

pguibert6WIND replied:
got it. my comment was about the comit log title prefix;

pguibert6WIND commented:
> > I tested it and this does not change the behaviour.
> > the issue takes place on hub, when receiving OSPF packets.
> 
> We've tested with FRR as hub also. Can you please tell what header check is failing? Do you get "ospf_read network address is not same" error also or just the "Header check failed"?

```
2021/03/18 09:03:27 OSPF: ospf_recv_packet: fd 15(vrf0) on interface 9(gre1)
2021/03/18 09:03:27 OSPF: [EC 134217741] interface gre1:11.255.255.1: ospf_read network address is not same [11.255.255.2]
2021/03/18 09:03:27 OSPF: ospf_read[11.255.255.2]: Header check failed, dropping.
```

amollad commented:
> > > I tested it and this does not change the behaviour.
> > > the issue takes place on hub, when receiving OSPF packets.
> > 
> > 
> > We've tested with FRR as hub also. Can you please tell what header check is failing? Do you get "ospf_read network address is not same" error also or just the "Header check failed"?
> 
> ```
> 2021/03/18 09:03:27 OSPF: ospf_recv_packet: fd 15(vrf0) on interface 9(gre1)
> 2021/03/18 09:03:27 OSPF: [EC 134217741] interface gre1:11.255.255.1: ospf_read network address is not same [11.255.255.2]
> 2021/03/18 09:03:27 OSPF: ospf_read[11.255.255.2]: Header check failed, dropping.
> ```

Below code is added in the patch to address the above error.  I assume you have configured ip address of gre1 on both ends (hub and spoke) with /32 prefix? If yes, can you please add a debug and see why ospf_check_network_mask() is returning error? Sorry for this, but as you can see it's a simple logic and I'm not sure why it's failing at your end.

ospf_packet.c:
static int ospf_check_network_mask(..) {
....
	/* Ignore mask check for max prefix length (32) */
	if (oi->type == OSPF_IFTYPE_POINTOMULTIPOINT
	    && oi->address->prefixlen == IPV4_MAX_BITLEN)
		return 1;
....
}

pguibert6WIND commented:
> > > > I tested it and this does not change the behaviour.
> > > > the issue takes place on hub, when receiving OSPF packets.
> > > 
> > > 
> > > We've tested with FRR as hub also. Can you please tell what header check is failing? Do you get "ospf_read network address is not same" error also or just the "Header check failed"?
> > 
> > 
> > ```
> > 2021/03/18 09:03:27 OSPF: ospf_recv_packet: fd 15(vrf0) on interface 9(gre1)
> > 2021/03/18 09:03:27 OSPF: [EC 134217741] interface gre1:11.255.255.1: ospf_read network address is not same [11.255.255.2]
> > 2021/03/18 09:03:27 OSPF: ospf_read[11.255.255.2]: Header check failed, dropping.
> > ```
> 
> Below code is added in the patch to address the above error. I assume you have configured ip address of gre1 on both ends (hub and spoke) with /32 prefix? If yes, can you please add a debug and see why ospf_check_network_mask() is returning error? Sorry for this, but as you can see it's a simple logic and I'm not sure why it's failing at your end.
> 
> ospf_packet.c:
> static int ospf_check_network_mask(..) {
> ....
> /* Ignore mask check for max prefix length (32) */
> if (oi->type == OSPF_IFTYPE_POINTOMULTIPOINT
> && oi->address->prefixlen == IPV4_MAX_BITLEN)
> return 1;
> ....
> }

ok, the hub config had disappeared on my setup.
this is ok, now functionality is good.
thanks for feedback.
I will approve the pull request once the minor fixes are reworked.
thanks.

reubendowle commented on nhrpd/linux.c:
```diff
@@ -32,6 +32,11 @@
+size_t strlcpy(char *__restrict dest,
+               const char *__restrict src, size_t destsize);
+#endif
+
```
> 
> this can be avoided by including "zebra.h" instead.
> but you can keep it as is for now.

Code does not actually compile when zebra.h is included, because zebra definitions for a bunch of things clash with the values brought in from the kernel header files. Resolving all that seemed not worth the effort, so I just put the forward declaration for the function there directly.

pguibert6WIND replied:
this is fine. thanks for having tried.

reubendowle commented on ospfd/ospf_spf.c:
```diff
@@ -836,6 +836,7 @@ static unsigned int ospf_nexthop_calculation(struct ospf_area *area,
 					oi = ospf_if_lookup_by_lsa_pos(area,
 								       lsa_pos);
+
 					if (!oi) {
```
done

reubendowle commented on ospfd/ospf_packet.c:
```diff
@@ -907,8 +907,11 @@ static void ospf_hello(struct ip *iph, struct ospf_header *ospfh,
 	if (oi->type != OSPF_IFTYPE_POINTOPOINT
-	    && oi->type != OSPF_IFTYPE_VIRTUALLINK)
+	    && oi->type != OSPF_IFTYPE_VIRTUALLINK
+	    && !(oi->type == OSPF_IFTYPE_POINTOMULTIPOINT
```
done

pguibert6WIND commented:
almost ok.
just 2 changes to be done, and this will ok, please.
```
        debugf(NHRP_DEBUG_COMMON, "Received multicast packet on %s len %zu\n",
	           ctx.ifp->name, zbuf_used(ctx.pkt));
```


-    \n is not necessary.
-    also the comment may be more explicit. I did not realize at first, but we need to understand that this is a packet going out that is intercepted ( and not incoming ospf messages).
    what about message like:
    "Intercepted multicast packet leaving %s len %zu"

mwinter-osr commented:
ci:rerun

pguibert6WIND reviewed (approved):
LGTM.
please rebase; test may probably fail because of that.

reubendowle commented:
> 
> LGTM.
> please rebase; test may probably fail because of that.

@pguibert6WIND I rebased and pushed again, and now it has failed with some other seemingly random errors.

pguibert6WIND commented:
> > LGTM.
> > please rebase; test may probably fail because of that.
> 
> @pguibert6WIND I rebased and pushed again, and now it has failed with some other seemingly random errors.

at least compilation error on fedora or centos
```
error	23-Mar-2021 20:32:50	nhrpd/nhrp_multicast.c:33:1: error: expected ';' before 'int'
error	23-Mar-2021 20:32:50	 int netlink_mcast_nflog_group;
error	23-Mar-2021 20:32:50	 ^
```

an other warning related to documentation, perhaps not fatal, but may have been introduced by a separate pr
```
error	23-Mar-2021 20:34:10	/home/ci/cibuild.17942/rpmwork/rpmbuild/BUILD/frr-7.7-dev-20210324-git.f87bd0f/doc/user/nhrpd.rst:250: WARNING: image file not readable: ../figures/fig_dmvpn_topologies.png
```

reubendowle commented:
> 
> 
> > > LGTM.
> > > please rebase; test may probably fail because of that.
> > 
> > 
> > @pguibert6WIND I rebased and pushed again, and now it has failed with some other seemingly random errors.
> 
> at least compilation error on fedora or centos
> 
I was 

> ```
> error	23-Mar-2021 20:32:50	nhrpd/nhrp_multicast.c:33:1: error: expected ';' before 'int'
> error	23-Mar-2021 20:32:50	 int netlink_mcast_nflog_group;
> error	23-Mar-2021 20:32:50	 ^
> ```
> 
> an other warning related to documentation, perhaps not fatal, but may have been introduced by a separate pr
> 
> ```
> error	23-Mar-2021 20:34:10	/home/ci/cibuild.17942/rpmwork/rpmbuild/BUILD/frr-7.7-dev-20210324-git.f87bd0f/doc/user/nhrpd.rst:250: WARNING: image file not readable: ../figures/fig_dmvpn_topologies.png
> ```



> 
> 
> > > LGTM.
> > > please rebase; test may probably fail because of that.
> > 
> > 
> > @pguibert6WIND I rebased and pushed again, and now it has failed with some other seemingly random errors.
> 
> at least compilation error on fedora or centos
> 

I was just looking at the NetDef-CI post above which is all garbled. But yes looking at the actual detailed report I can see this.

> ```
> error	23-Mar-2021 20:32:50	nhrpd/nhrp_multicast.c:33:1: error: expected ';' before 'int'
> error	23-Mar-2021 20:32:50	 int netlink_mcast_nflog_group;
> error	23-Mar-2021 20:32:50	 ^
> ```
> 

No idea how this missing semicolon got in there. It seems to compile ok on most platforms. Perhaps the referenced macro changed to now need a semicolon when I rebased? Anyway, it is now fixed.

> an other warning related to documentation, perhaps not fatal, but may have been introduced by a separate pr
> 
> ```
> error	23-Mar-2021 20:34:10	/home/ci/cibuild.17942/rpmwork/rpmbuild/BUILD/frr-7.7-dev-20210324-git.f87bd0f/doc/user/nhrpd.rst:250: WARNING: image file not readable: ../figures/fig_dmvpn_topologies.png
> ```

This seems to be a bug in the build system. The file is present, and it certainly looks like a perfectly normal png when I open it. This was all added in a commit from months ago.

mwinter-osr commented:
So the key part to remember here is that the RPM packages are built from the output of the "make dist" and no other sources. If someone adds a new file, but somehow forgets to add it to the dist target, then it will be missing during the RPM build (but the other builds still work fine as they are based on git)

reubendowle commented:
> 
> So the key part to remember here is that the RPM packages are built from the output of the "make dist" and no other sources. If someone adds a new file, but somehow forgets to add it to the dist target, then it will be missing during the RPM build (but the other builds still work fine as they are based on git)

I made no changes for this (I just fixed other missing semicolon issue), but it built fine. So this must be something unreliable in the build system.

pguibert6WIND commented:
> > So the key part to remember here is that the RPM packages are built from the output of the "make dist" and no other sources. If someone adds a new file, but somehow forgets to add it to the dist target, then it will be missing during the RPM build (but the other builds still work fine as they are based on git)
> 
> I made no changes for this (I just fixed other missing semicolon issue), but it built fine. So this must be something unreliable in the build system.

I think doc/subdir.am should be updated.
```
guibert@yngwie:~/DOCKER/frr3$ git grep git_branches.svg
doc/figures/git_branches.svg:   sodipodi:docname="git_branches.svg">
doc/subdir.am:  doc/figures/git_branches.svg \
guibert@yngwie:~/DOCKER/frr3$ git grep fig_dmvpn_topologies
doc/user/nhrpd.rst:.. figure:: ../figures/fig_dmvpn_topologies.png
```

reubendowle commented:
> 
> 
> > > So the key part to remember here is that the RPM packages are built from the output of the "make dist" and no other sources. If someone adds a new file, but somehow forgets to add it to the dist target, then it will be missing during the RPM build (but the other builds still work fine as they are based on git)
> > 
> > 
> > I made no changes for this (I just fixed other missing semicolon issue), but it built fine. So this must be something unreliable in the build system.
> 
> I think doc/subdir.am should be updated.
> 
> ```
> guibert@yngwie:~/DOCKER/frr3$ git grep git_branches.svg
> doc/figures/git_branches.svg:   sodipodi:docname="git_branches.svg">
> doc/subdir.am:  doc/figures/git_branches.svg \
> guibert@yngwie:~/DOCKER/frr3$ git grep fig_dmvpn_topologies
> doc/user/nhrpd.rst:.. figure:: ../figures/fig_dmvpn_topologies.png
> ```

That sounds reasonable. But I don't think that change belongs in this pull-request, as that is an unrelated change.

Jafaral commented:
ci:rerun

Jafaral reviewed (approved):
LGTM at this point!

Jafaral commented:
@reubendowle  minor styling issues remaining.

reubendowle commented:
> 
> @reubendowle minor styling issues remaining.

I have fixed some of the style issues, but the issue on line 909 and 910 of ospf_spf.c I have left as-is. The line that is being replaced was already exceeding the 80 character per line rule, and I don't see any good way to fix the issue. Clang-format recommends something that looks really mangled, and yet still exceeds 80 characters. So unless you can suggest something better I prefer it like it currently is.

pguibert6WIND commented:
ci:rerun

Jafaral commented on ospfd/ospf_spf.c:
```diff
@@ -905,7 +905,9 @@ static unsigned int ospf_nexthop_calculation(struct ospf_area *area,
-				    || (oi && oi->type == OSPF_IFTYPE_POINTOPOINT)) {
+				    || (oi && oi->type == OSPF_IFTYPE_POINTOPOINT)
+				    || (oi && oi->type == OSPF_IFTYPE_POINTOMULTIPOINT
+					   && oi->address->prefixlen == IPV4_MAX_BITLEN)) {
```
@qlyoung do you know a  good fix for this warning? 

see https://github.com/FRRouting/frr/pull/8153#issuecomment-811573353

qlyoung replied:
@polychaeta said nothing after the fix, I'm not sure what warning you're referring to

Jafaral replied:
@qlyoung , thanks. Just wanted to make sure that If I merge this PR the warning wont trickle down to allow new PRs.

Jafaral commented:
ci:rerun

Jafaral commented on nhrpd/nhrp_multicast.c:
```diff
@@ -0,0 +1,310 @@
+	debugf(NHRP_DEBUG_COMMON,
+	       "Multicast Packet: %s -> %s, ret = %d, size = %zu, addrlen = %zu",
+	       sockunion2str(&p->vc->local.nbma, buf[0], sizeof(buf[0])),
+	       sockunion2str(&p->vc->remote.nbma, buf[1], sizeof(buf[1])), ret,
```
@reubendowle, I hate to keep coming back to ask you make changes, but sockunion2str is being dropped in favor of %pSU see #8411. Can you please fix these ?

Jafaral commented on nhrpd/nhrp_multicast.c:
```diff
@@ -0,0 +1,310 @@
+	};
+	list_add_tail(&mcast->list_entry, &nifp->afi[afi].mcastlist_head);
+
+	sockunion2str(nbma_addr, buf, sizeof(buf));
```
same comment above, sockunion2str ==> %pSU

Jafaral commented on nhrpd/nhrp_multicast.c:
```diff
@@ -0,0 +1,310 @@
+		if (!sockunion_same(&mcast->nbma_addr, nbma_addr))
+			continue;
+
+		sockunion2str(nbma_addr, buf, sizeof(buf));
```
same comment above, sockunion2str ==> %pSU

Jafaral commented on nhrpd/nhrp_vty.c:
```diff
@@ -732,8 +807,8 @@ static void show_ip_nhrp_nhs(struct nhrp_nhs *n, struct nhrp_registration *reg,
-		sockunion2str(&reg->peer->vc->remote.nbma,
-			      buf[0], sizeof(buf[0]));
+		sockunion2str(&reg->peer->vc->remote.nbma, buf[0],
+			      sizeof(buf[0]));
```
same comment above, sockunion2str ==> %pSU.

reubendowle replied:
In this case the string is used when outputting to json, so %pSU cannot be used here

Jafaral commented on nhrpd/nhrp_vty.c:
```diff
@@ -1059,7 +1135,8 @@ static void interface_config_write_nhrp_map(struct nhrp_cache_config *c, void *d
 		c->type == NHRP_CACHE_LOCAL
-		? "local" : sockunion2str(&c->nbma, buf[1], sizeof(buf[1])));
+			? "local"
+			: sockunion2str(&c->nbma, buf[1], sizeof(buf[1])));
```
same comment above, sockunion2str ==> %pSU

reubendowle replied:
Here the output can either be a fixed string ("local") or the sockunion contents. So %pSU won't work here.

Jafaral reviewed (commented):
@reubendowle  While adding these new comments I realized that that sockunionstr use is spread all over the place in nhrp. If you have bandwidth to make the change for the ones  you added please do. If you want to take a pass that would be OK too. I will do a round of cleanup for those anyway. Please let me know.

Jafaral commented:
ci:rerun

Jafaral commented:
ci:rerun
